src/merklized_state.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the HotShot Query Service library.
+//
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not,
+// see <https://www.gnu.org/licenses/>.
+
+//! Api for querying merklized state
+//!
+//! The state API provides an interface for serving queries against arbitrarily old snapshots of the state.
+//! This allows a full Merkle tree to be reconstructed from storage.
+//! If any parent state is missing then the partial snapshot can not be queried.
+use std::{fmt::Display, path::PathBuf};
+
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use clap::Args;
+use derive_more::From;
+use futures::FutureExt;
+use hotshot_types::traits::node_implementation::NodeType;
+use jf_primitives::merkle_tree::{Index, MerkleTreeScheme, ToTraversalPath};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use snafu::Snafu;
+use tagged_base64::TaggedBase64;
+use tide_disco::{api::ApiError, method::ReadState, Api, RequestError, StatusCode};
+
+use crate::api::load_api;
+
+pub(crate) mod data_source;
+pub use data_source::*;
+
+#[derive(Args, Default)]
+pub struct Options {
+    #[arg(
+        long = "merklized-state-api-path",
+        env = "HOTSHOT_MERKLIZED_STATE_API_PATH"
+    )]
+    pub api_path: Option<PathBuf>,
+
+    /// Additional API specification files to merge with `merklized-state-api-path`.
+    ///
+    /// These optional files may contain route definitions for application-specific routes that have
+    /// been added as extensions to the basic status API.
+    #[arg(
+        long = "merklized-state-extension",
+        env = "HOTSHOT_MERKLIZED_STATE_EXTENSIONS",
+        value_delimiter = ','
+    )]
+    pub extensions: Vec<toml::Value>,
+}
+
+#[derive(Clone, Debug, From, Snafu, Deserialize, Serialize)]
+pub enum Error {
```

Comment:
We should also have a variant wrapping `QueryError`, whose `status()` is `source.status()`. That allows us to return 404 e.g. if a query fails because some data was missing, rather than 500